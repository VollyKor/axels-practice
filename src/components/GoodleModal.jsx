import { useState, useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { GoogleMap, useLoadScript, Marker } from '@react-google-maps/api';
import Modal from 'react-bootstrap/Modal';
import Button from 'react-bootstrap/Button';

import { setGeo } from '../redux/ducks/geo';

const containerStyle = {
    width: '100%',
    height: '100%',

    minWidth: '200px',
    minHeight: '300px',
};

const options = {
    disableDefaultUI: true,
    zoomControl: true,
};

const libraries = ['places'];

const GoogleModal = ({ fillForm, ...props }) => {
    const [marker, setMarker] = useState(() => ({
        lat: 0,
        lng: 0,
    }));

    const geo = useSelector((state) => state.geo);
    const dispatch = useDispatch();

    useEffect(() => {
        setMarker({ ...geo });
    }, [dispatch, geo]);

    const { isLoaded, loadError } = useLoadScript({
        googleMapsApiKey: process.env.REACT_APP_GOOGLE_API_KEY,
        libraries,
        language: 'en',
    });

    if (loadError) return 'Google Map Error';
    if (!isLoaded) return 'Google Map Loading';

    return (
        <Modal
            {...props}
            size="lg"
            aria-labelledby="contained-modal-title-vcenter"
            centered
        >
            <Modal.Header closeButton>
                <Modal.Title id="contained-modal-title-vcenter">
                    Choose Your Address
                </Modal.Title>
            </Modal.Header>
            <Modal.Body>
                <GoogleMap
                    mapContainerStyle={containerStyle}
                    center={geo}
                    zoom={14}
                    options={options}
                    onClick={(event) => {
                        setMarker({
                            lat: event.latLng.lat(),
                            lng: event.latLng.lng(),
                        });
                    }}
                >
                    <Marker position={geo && marker} />
                </GoogleMap>
            </Modal.Body>
            <Modal.Footer>
                <Button
                    onClick={() => {
                        dispatch(setGeo(marker));
                        props.onHide();
                        fillForm(marker);
                    }}
                >
                    Submit
                </Button>
                <Button
                    variant="secondary"
                    onClick={() => {
                        props.onHide();
                    }}
                >
                    Close
                </Button>
            </Modal.Footer>
        </Modal>
    );
};

export default GoogleModal;
