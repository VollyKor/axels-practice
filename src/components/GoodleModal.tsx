import { useState, useEffect } from 'react';

import { useDispatch, useSelector } from 'react-redux';
import { GoogleMap, useLoadScript, Marker } from '@react-google-maps/api';
import Modal from 'react-bootstrap/Modal';
import Button from 'react-bootstrap/Button';

import { setGeo } from '../redux/ducks/geo';
import { Libraries } from '@react-google-maps/api/dist/utils/make-load-script-url';

const GoogleMapsStyles = {
    width: '100%',
    height: '100%',

    minWidth: '200px',
    minHeight: '300px',
};

const options = {
    disableDefaultUI: true,
    zoomControl: true,
};

const libraries: Libraries = ['places'];

interface Props {
    fillForm: TfillForm;
    onHide: Function;
    show: boolean;
}

const GoogleModal = ({ fillForm, ...props }: Props) => {
    const googleKey = process.env.REACT_APP_GOOGLE_API_KEY || 'unvalidKey';

    const [marker, setMarker] = useState(() => ({
        lat: 0,
        lng: 0,
    }));

    const geo = useSelector(({ geo }: RootState) => geo);
    const dispatch = useDispatch();

    useEffect(() => {
        setMarker({ ...geo });
    }, [dispatch, geo]);

    const { loadError } = useLoadScript({
        googleMapsApiKey: googleKey,
        libraries,
        language: 'en',
    });

    if (loadError) return <p>Google Map Error</p>;

    return (
        <Modal
            {...props}
            size="lg"
            aria-labelledby="contained-modal-title-vcenter"
            centered
        >
            <Modal.Header closeButton>
                <Modal.Title id="contained-modal-title-vcenter">
                    Choose Your Address
                </Modal.Title>
            </Modal.Header>

            <Modal.Body>
                <GoogleMap
                    mapContainerStyle={GoogleMapsStyles}
                    center={geo}
                    zoom={14}
                    options={options}
                    onClick={(event) => {
                        if (event.latLng && event.latLng) {
                            setMarker({
                                lat: event.latLng.lat(),
                                lng: event.latLng.lng(),
                            });
                        }
                    }}
                >
                    <Marker position={geo && marker} />
                </GoogleMap>
            </Modal.Body>

            <Modal.Footer>
                <Button
                    onClick={() => {
                        dispatch(setGeo(marker));
                        props.onHide();
                        fillForm(marker);
                    }}
                >
                    Submit
                </Button>
                <Button
                    variant="secondary"
                    onClick={() => {
                        props.onHide();
                    }}
                >
                    Close
                </Button>
            </Modal.Footer>
        </Modal>
    );
};

export default GoogleModal;
