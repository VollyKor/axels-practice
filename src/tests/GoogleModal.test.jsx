import { shallow } from 'enzyme';
import * as redux from 'react-redux';
import { GoogleMap } from '@react-google-maps/api';

import GoodleModal from 'components/GoodleModal';

describe('test suite', () => {
    let spyOnUseSelector;
    let spyOnUseDispatch;
    let mockDispatch;

    const ititialCoords = { lat: 21, lng: 32 };

    beforeEach(() => {
        // Mock useSelector hook
        spyOnUseSelector = jest.spyOn(redux, 'useSelector');
        spyOnUseSelector.mockReturnValue({
            lat: 21,
            lng: 32,
        });

        // Mock useDispatch hook
        spyOnUseDispatch = jest.spyOn(redux, 'useDispatch');
        // Mock dispatch function returned from useDispatch
        mockDispatch = jest.fn();
        spyOnUseDispatch.mockReturnValue(mockDispatch);
    });

    afterEach(() => {
        jest.restoreAllMocks();
    });

    it('should render', () => {
        const wrapper = shallow(<GoodleModal />);
        expect(wrapper.exists()).toBe(true);
    });

    it('renders correctly', () => {
        const wrapper = shallow(<GoodleModal />);
        expect(wrapper).toMatchSnapshot();
    });

    it('change markers', () => {
        const wrapper = shallow(<GoodleModal />);

        setTimeout(() => {
            wrapper.update();
            const map = wrapper.find(GoogleMap);

            expect(map.prop('center')).toEqual(ititialCoords);
        });
    });
});
