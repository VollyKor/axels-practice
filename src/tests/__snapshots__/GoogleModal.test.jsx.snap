// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test suite renders correctly 1`] = `"Google Map Loading"`;
